### ISSUE TITLE
JVM Settings

### ISSUE THREAD

**Issue by qrpike** [Yolean/kubernetes-kafka#112]
Since using a JVM in containers is already somewhat scary, setting limits on the pods is always a good idea, however there are no mentions of this in the docs. 

Adding a:
```
        env:
        - name: "KAFKA_HEAP_OPTS"
          value: "-Xmx6G -Xms6G"
        resources:
          limits:
            cpu: 4
            memory: 8G
```

In the docs may be a good idea. JVMs tend to consume memory unless told otherwise.

**Comment by solsson** [comment]
> Since using a JVM in containers is already somewhat scary

agreed :)

> In the docs

Do you mean in the readme? I think it will always have to be tweaked per installation based on the load etc. In https://github.com/Yolean/kubernetes-kafka/pull/44#issuecomment-347035123 I'm looking at how to minimize the footprint, for testing.

**Comment by StevenACoffman** [comment]
I'm actually pretty interested in this topic (memory and cpu requests/limits for java apps in kubernetes) and very interested in your opinions. The best advice I've found is jsravn [from his experience building a kubernetes stack from scratch](https://jsravn.com/2017/07/15/a-kubernetes-stack-from-scratch/). He says:
+ `-Xmx` to roughly half the size of the memory limits of the pod. There is a lot of memory overhead in hotspot. This value will depend on the application as well, so it takes some work to find the right amount of max heap to prevent an OOM kill.
+ `-XX:+AlwaysPreTouch` so we can more quickly ensure our max heap setting is reasonable. Otherwise we might only get OOM killed under load.
+ Never limit CPU on a JVM container, unless you really don’t care about long pause times. The CPU limit will cause hard throttling of your application whenever garbage collections happen. The right choice almost always is to use burstable CPU. This let’s the scheduler soft clamp the application if it uses too much CPU in relation to other applications running on the node, while also letting it use as much CPU as available on the machine.

**Comment by solsson** [comment]
Great input. The first task I had to battle with was to select a JVM vendor+version. Am I wrong to assume that the choice there affects behavior with memory flags?

[The default kafka image here](https://github.com/solsson/dockerfiles/tree/master/jre-zulu) uses Zulu because [Confluent](https://github.com/confluentinc/cp-docker-images/issues/21) does so. I'm obviously not a lawyer but I think that anyone who builds their own images can use Oracle's server-jre.

Are there other options? Is openjdk:8-jre-slim a better option than Zulu? Or should we go straight for Java9 before we start tweaking, given the improvement in kafka SSL performance? I imagine that the java community has efforts underway to tweak JVMs for use in Docker, and even with Kubernetes memory limits. Haven't researched that topic though.

Now back to the JVM args... I battled with them around https://github.com/Yolean/kubernetes-kafka/pull/49#issuecomment-319174644 and was quite unhappy about the outcome (hadn't seen jsravn's advice at that time). After that I didn't dare to merge anything with memory limits on broker pods. But I'm willing to un-conclude that :)

**Comment by StevenACoffman** [comment]
With regards to Kafka, doing whatever Confluent does is generally the best advice. Sadly, probably best to avoid Oracle for legal reasons and pick Zulu openJDK8. JDK 9 shows great promise, and is fun for experiments, and maybe individual apps in production, but I would not run critical infrastructure like kafka in prod on it until Confluent does.

I can only speak to tuning web apps in java in kubernetes, and I worry that Kafka is different enough to invalidate that experience.

**Comment by solsson** [comment]
Haven't read it yet, but https://dzone.com/articles/why-my-java-application-is-oomkilled could be relevant.

**Comment by solsson** [comment]
The results of #128 will be very interesting. There, for the `metrics` container I've only enabled memory limit awareness. No other flags that affect memory use.

I'll keep a close watch on the Prometheus metric `container_memory_usage_bytes{namespace="kafka",container_name="metrics"}`, to get a baseline for experiments with memory flags later.

**Comment by qrpike** [comment]
@solsson You have a prometheus exporter for the image now? I had to extend you image to add one, pretty important in my setup. Thanks

**Comment by solsson** [comment]
I didn't add the exporter to the image. It's a separate container in the pod, added using kubectl patch, see #128. Did you find an exporter that plugins into kafka without JMX? I did a couple of experiments based on https://github.com/arnobroekhof/kafka-http-metrics-reporter,  https://github.com/prepor/kafka-prometheus and https://github.com/prometheus/client_java/blob/master/simpleclient_dropwizard but I failed to get any useful results.

**Comment by qrpike** [comment]
@solsson  No, I currently use the JMX exporter. Is there a downside to using the JMX exporter? I'm not hugely experienced with the JVM.

Thanks,

On a side note, some of my consumers seem to lag behind when there is a lot of data ( 20k/sec ). However they aren't using much cpu or network. Obviously consumer lib etc could affect this, but on the actual Kafka side is there anything I should be monitoring to possible see whats slowing it down?

Thanks x 2

**Comment by solsson** [comment]
> Is there a downside to using the JMX exporter?

@qrpike Ok, I think we're all ending up with JMX as the only viable alternative. As a sidecar it's relatively resource heavy, and before #128 I had failed to set a reasonable memory limit that didn't affect stability of the kafka pod.

 > some of my consumers seem to lag behind when there is a lot of data ( 20k/sec ). However they aren't using much cpu or network. Obviously consumer lib etc could affect this, but on the actual Kafka side is there anything I should be monitoring to possible see whats slowing it down?

This is a great topic and quite core to Kafka ops. I'm no expert but I'd like to see the community discuss this. I suggest you open a new issue with the stats you have. Please clarify if it is the cpu & network of consuming pods that is low. Also what metrics you have from Kafka pods and nodes.

**Comment by solsson** [comment]
See f9162a6 and don't hesitate to open new discussions.
